File: day06/zig/src/main.zig
const std = @import("std");

const Problem = struct {
    numbers: std.ArrayList(u128),
    operation: u8,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var stdout_buf: [256]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buf);
    const stdout = &stdout_writer.interface;

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var lines: std.ArrayList([]const u8) = .empty;
    defer lines.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        if (line.len > 0) {
            try lines.append(allocator, line);
        }
    }

    // Pad lines to max width
    var max_width: usize = 0;
    for (lines.items) |line| {
        if (line.len > max_width) max_width = line.len;
    }

    var padded_lines: std.ArrayList([]u8) = .empty;
    defer {
        for (padded_lines.items) |line| allocator.free(line);
        padded_lines.deinit(allocator);
    }

    for (lines.items) |line| {
        var padded = try allocator.alloc(u8, max_width);
        @memcpy(padded[0..line.len], line);
        @memset(padded[line.len..], ' ');
        try padded_lines.append(allocator, padded);
    }

    // Part 1: Row-wise parsing
    var problems_part1 = try extractProblems(allocator, padded_lines.items, parseRowWise);
    defer {
        for (problems_part1.items) |*p| p.numbers.deinit(allocator);
        problems_part1.deinit(allocator);
    }

    var grand_total_part1: u128 = 0;
    for (problems_part1.items) |problem| {
        grand_total_part1 += evaluateProblem(problem);
    }

    // Part 2: Column-wise parsing
    var problems_part2 = try extractProblems(allocator, padded_lines.items, parseColumnWise);
    defer {
        for (problems_part2.items) |*p| p.numbers.deinit(allocator);
        problems_part2.deinit(allocator);
    }

    var grand_total_part2: u128 = 0;
    for (problems_part2.items) |problem| {
        grand_total_part2 += evaluateProblem(problem);
    }

    try stdout.print("Part 1: {d}\n", .{grand_total_part1});
    try stdout.print("Part 2: {d}\n", .{grand_total_part2});
    try stdout.flush();
}

fn isBlankColumn(lines: []const []u8, col: usize) bool {
    for (lines) |line| {
        if (col < line.len and line[col] != ' ') {
            return false;
        }
    }
    return true;
}

const ColumnRange = struct { start: usize, end: usize };
const ParseFn = *const fn (allocator: std.mem.Allocator, lines: []const []u8, start_col: usize, end_col: usize) anyerror!Problem;

fn extractProblems(allocator: std.mem.Allocator, lines: []const []u8, parser: ParseFn) !std.ArrayList(Problem) {
    const width = if (lines.len > 0) lines[0].len else 0;

    var column_ranges: std.ArrayList(ColumnRange) = .empty;
    defer column_ranges.deinit(allocator);

    var block_start: ?usize = null;
    for (0..width) |col| {
        const is_blank = isBlankColumn(lines, col);
        if (block_start == null and !is_blank) {
            block_start = col;
        } else if (block_start != null and is_blank) {
            try column_ranges.append(allocator, .{ .start = block_start.?, .end = col });
            block_start = null;
        }
    }
    if (block_start) |start| {
        try column_ranges.append(allocator, .{ .start = start, .end = width });
    }

    var problems: std.ArrayList(Problem) = .empty;
    for (column_ranges.items) |range| {
        try problems.append(allocator, try parser(allocator, lines, range.start, range.end));
    }

    return problems;
}

fn parseRowWise(allocator: std.mem.Allocator, lines: []const []u8, start_col: usize, end_col: usize) !Problem {
    var numbers: std.ArrayList(u128) = .empty;
    var operation: u8 = '+';

    for (lines) |line| {
        const end = @min(end_col, line.len);
        if (start_col >= end) continue;

        const row_slice = line[start_col..end];

        // Extract digits
        var digits_buf: [64]u8 = undefined;
        var digits_len: usize = 0;
        for (row_slice) |c| {
            if (std.ascii.isDigit(c)) {
                digits_buf[digits_len] = c;
                digits_len += 1;
            }
        }

        if (digits_len > 0) {
            const num = try std.fmt.parseInt(u128, digits_buf[0..digits_len], 10);
            try numbers.append(allocator, num);
        }

        // Find operation
        for (row_slice) |c| {
            if (c == '+' or c == '*') {
                operation = c;
                break;
            }
        }
    }

    return Problem{ .numbers = numbers, .operation = operation };
}

fn parseColumnWise(allocator: std.mem.Allocator, lines: []const []u8, start_col: usize, end_col: usize) !Problem {
    var numbers: std.ArrayList(u128) = .empty;
    var operation: u8 = '+';

    // Process columns right-to-left
    var col = end_col;
    while (col > start_col) {
        col -= 1;

        // Extract column chars
        var digits_buf: [64]u8 = undefined;
        var digits_len: usize = 0;
        var found_op: ?u8 = null;

        for (lines) |line| {
            const c = if (col < line.len) line[col] else ' ';
            if (std.ascii.isDigit(c)) {
                digits_buf[digits_len] = c;
                digits_len += 1;
            }
            if (c == '+' or c == '*') {
                found_op = c;
            }
        }

        if (digits_len > 0) {
            const num = try std.fmt.parseInt(u128, digits_buf[0..digits_len], 10);
            try numbers.append(allocator, num);
        }
        if (found_op) |op| {
            operation = op;
        }
    }

    // Reverse to get left-to-right order
    std.mem.reverse(u128, numbers.items);

    return Problem{ .numbers = numbers, .operation = operation };
}

fn evaluateProblem(problem: Problem) u128 {
    if (problem.numbers.items.len == 0) return 0;

    if (problem.operation == '+') {
        var sum: u128 = 0;
        for (problem.numbers.items) |n| {
            sum += n;
        }
        return sum;
    } else {
        var product: u128 = 1;
        for (problem.numbers.items) |n| {
            product *= n;
        }
        return product;
    }
}

File: day06/rust/src/main.rs
use num_bigint::BigInt;
use std::fs;

struct Problem {
    numbers: Vec<BigInt>,
    operation: char,
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let lines: Vec<&str> = input.lines().filter(|l| !l.is_empty()).collect();

    let max_width = lines.iter().map(|l| l.len()).max().unwrap_or(0);
    let padded_lines: Vec<String> = lines
        .iter()
        .map(|l| format!("{:width$}", l, width = max_width))
        .collect();
    let padded_refs: Vec<&str> = padded_lines.iter().map(|s| s.as_str()).collect();

    let problems_part1 = extract_problems(&padded_refs, parse_row_wise);
    let grand_total_part1: BigInt = problems_part1.iter().map(evaluate_problem).sum();
    println!("Part 1: {}", grand_total_part1);

    let problems_part2 = extract_problems(&padded_refs, parse_column_wise);
    let grand_total_part2: BigInt = problems_part2.iter().map(evaluate_problem).sum();
    println!("Part 2: {}", grand_total_part2);
}

fn is_blank_column(lines: &[&str], col: usize) -> bool {
    lines.iter().all(|line| {
        col >= line.len() || line.chars().nth(col).unwrap_or(' ') == ' '
    })
}

fn extract_problems<F>(lines: &[&str], parser: F) -> Vec<Problem>
where
    F: Fn(&[&str], usize, usize) -> Problem,
{
    let width = lines[0].len();
    let mut column_ranges: Vec<(usize, usize)> = Vec::new();
    let mut block_start: Option<usize> = None;

    for col in 0..width {
        let is_blank = is_blank_column(lines, col);
        match (block_start, is_blank) {
            (None, false) => block_start = Some(col),
            (Some(start), true) => {
                column_ranges.push((start, col));
                block_start = None;
            }
            _ => {}
        }
    }

    if let Some(start) = block_start {
        column_ranges.push((start, width));
    }

    column_ranges
        .iter()
        .map(|&(start, end)| parser(lines, start, end))
        .collect()
}

fn parse_row_wise(lines: &[&str], start_col: usize, end_col: usize) -> Problem {
    let mut numbers: Vec<BigInt> = Vec::new();
    let mut operation = '+';

    for line in lines {
        let end = end_col.min(line.len());
        if start_col >= end {
            continue;
        }
        let row_slice = &line[start_col..end];
        let digits: String = row_slice.chars().filter(|c| c.is_ascii_digit()).collect();
        let found_op = row_slice.chars().find(|&c| c == '+' || c == '*');

        if !digits.is_empty() {
            numbers.push(digits.parse().unwrap());
        }
        if let Some(op) = found_op {
            operation = op;
        }
    }

    Problem { numbers, operation }
}

fn parse_column_wise(lines: &[&str], start_col: usize, end_col: usize) -> Problem {
    let mut numbers: Vec<BigInt> = Vec::new();
    let mut operation = '+';

    for col in (start_col..end_col).rev() {
        let column_chars: Vec<char> = lines
            .iter()
            .map(|line| line.chars().nth(col).unwrap_or(' '))
            .collect();

        let digits: String = column_chars.iter().filter(|c| c.is_ascii_digit()).collect();
        let found_op = column_chars.iter().find(|&&c| c == '+' || c == '*');

        if !digits.is_empty() {
            numbers.push(digits.parse().unwrap());
        }
        if let Some(&op) = found_op {
            operation = op;
        }
    }

    numbers.reverse();
    Problem { numbers, operation }
}

fn evaluate_problem(problem: &Problem) -> BigInt {
    match problem.operation {
        '+' => problem.numbers.iter().cloned().sum(),
        '*' => problem.numbers.iter().cloned().product(),
        _ => BigInt::from(0),
    }
}
